File: DeepDive/arrayInterop.f90
module arrayInterop
  use iso_c_binding

type, bind(C) :: myType
  integer(c_int) :: year
  real(c_double) :: temperature(20)
end type myType

contains
  subroutine convertFtoC(myTypeF, myTypeC) bind(C, name="convertFtoC")
    type(myType), intent(in) :: myTypeF
    type(myType), intent(out) :: myTypeC

    myTypeC%year = myTypeF%year
    myTypeC%temperature = (myTypeF%temperature - 32.0) / 1.8
  end subroutine convertFtoC


end module arrayInterop

File: DeepDive/arrayInterop.c
#include <stdlib.h>
#include <stdio.h>
struct myType {
  int year;
  double temperature[20];
};

extern void convertFtoC(struct myType * myTypeF, struct myType * myTypeC);

int main() {
  struct myType myTypeF;
  struct myType myTypeC;

  myTypeF.year = 2025;
  for(int i = 0; i < 20; i++) {
    myTypeF.temperature[i] = 32 + 0.5*i;
  }

  convertFtoC(&myTypeF, &myTypeC);

  printf("Fahrenheit: ");
  for(int i=0 ; i < 20; i++) {
    printf("%f ", myTypeF.temperature[i]);
  }

  printf("\nCelsius: ");
  for(int i=0 ; i < 20; i++) {
    printf("%f ", myTypeC.temperature[i]);
  }
  printf("\n");
  return 0;
}